File: apps/api/src/observability.ts
import * as Sentry from "@sentry/node";
import type { AppConfig } from "./config.js";

/**
 * Error tracking. Sentry is initialised **only when SENTRY_DSN is set**, so
 * dev/test boot with no secret and the SDK stays dormant. When it is set, PII is
 * never sent and a scrubber redacts known-sensitive keys before any event
 * leaves the process. Structured request logging is already handled by Fastify's
 * pino logger (server.ts) — this is purely exception capture.
 */

let started = false;

const SCRUB_KEYS = new Set([
  "x-internal-secret",
  "authorization",
  "cookie",
  "set-cookie",
  "internal_shared_secret",
  "database_url",
  "redis_url",
  "cookie_secret",
  "master_encryption_key",
  "sentry_dsn",
  "password",
  "token",
  "secret",
  "credentials",
  "ciphertext",
  "encrypteddek",
]);

function redact(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(redact);
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([k, v]) => [
        k,
        SCRUB_KEYS.has(k.toLowerCase()) ? "[redacted]" : redact(v),
      ]),
    );
  }
  return value;
}

/** Wire Sentry iff a DSN is configured. Returns true when actually started. */
export function initSentry(config: AppConfig): boolean {
  if (started) return true;
  if (!config.SENTRY_DSN) return false;

  Sentry.init({
    dsn: config.SENTRY_DSN,
    environment: config.NODE_ENV,
    sendDefaultPii: false,
    tracesSampleRate: 0,
    beforeSend: (event) => redact(event) as Sentry.ErrorEvent,
  });
  started = true;
  return true;
}

/** Report an exception if Sentry is active; a no-op otherwise. */
export function captureException(err: unknown): void {
  if (started) Sentry.captureException(err);
}

File: services/pipeline/app/core/observability.py
"""Structured logging + optional Sentry for the pipeline.

Logging is stdlib-only (no new runtime dep): one configured root logger at the
settings level. We deliberately log only ids, counts, and stage/status — never
the Settings object, request bodies, scraped page content, model output, or
credentials. Sentry is initialised **only when SENTRY_DSN is set**, so mock/dev/
test runs are unaffected and the service boots with no secret required; when it is
set, a scrubber drops known-sensitive fields before any event leaves the process.
"""

from __future__ import annotations

import logging
from typing import Any

from app.core.config import get_settings

_LOG_CONFIGURED = False
_SENTRY_STARTED = False

# Header / field names that must never leave the process in an error report.
_SCRUB_KEYS = {
    "x-internal-secret",
    "authorization",
    "cookie",
    "set-cookie",
    "internal_shared_secret",
    "database_url",
    "redis_url",
    "openai_api_key",
    "anthropic_api_key",
    "google_api_key",
    "perplexity_api_key",
    "deepseek_api_key",
    "moonshot_api_key",
    "groq_api_key",
    "openrouter_api_key",
    "sentry_dsn",
    "password",
    "token",
    "secret",
    "ciphertext",
    "encrypted_dek",
}


def configure_logging() -> None:
    """Idempotently configure the root logger at the settings log level."""
    global _LOG_CONFIGURED
    if _LOG_CONFIGURED:
        return
    settings = get_settings()
    level = getattr(logging, settings.log_level.upper(), logging.INFO)
    logging.basicConfig(
        level=level,
        format="%(asctime)s %(levelname)s %(name)s %(message)s",
        force=True,
    )
    _LOG_CONFIGURED = True


def _scrub(event: dict[str, Any], _hint: dict[str, Any]) -> dict[str, Any]:
    """Sentry before_send: recursively redact known-sensitive keys."""

    def redact(obj: Any) -> Any:
        if isinstance(obj, dict):
            return {
                k: ("[redacted]" if k.lower() in _SCRUB_KEYS else redact(v))
                for k, v in obj.items()
            }
        if isinstance(obj, list):
            return [redact(v) for v in obj]
        return obj

    return redact(event)


def init_sentry() -> bool:
    """Wire Sentry iff SENTRY_DSN is set. Returns True when actually started.

    PII is never sent (`send_default_pii=False`) and a scrubber redacts secrets.
    Safe to call from both the FastAPI app and the Celery worker on boot."""
    global _SENTRY_STARTED
    if _SENTRY_STARTED:
        return True
    dsn = get_settings().sentry_dsn
    if not dsn:
        return False

    import sentry_sdk
    from sentry_sdk.integrations.celery import CeleryIntegration
    from sentry_sdk.integrations.fastapi import FastApiIntegration

    sentry_sdk.init(
        dsn=dsn,
        environment=get_settings().environment,
        integrations=[FastApiIntegration(), CeleryIntegration()],
        send_default_pii=False,
        traces_sample_rate=0.0,
        before_send=_scrub,
    )
    _SENTRY_STARTED = True
    return True
